### PR TITLE
Add ability to define and invoke methods (#487)

### PR DESCRIPTION
**Pull request by elias-winberg**
This pull request addresses issue https://github.com/amir20/phantomjs-node/issues/487, and it:

- adds `page.invokeAsyncMethod`, which invokes any asynchronous method ('includeJs', 'open', etc.)

- adds `page.invokeMethod`, which invokes any method

- adds `page.defineMethod`, which defines a method

- reimplements `property`, `setting` as commands rather than methods 

- binds the built-in/native methods to `invokeAsyncMethod`, and `invokeMethod`

- removes some deprecated code

Examples:

```js
'use strict';

const phantomjs = require('phantom');

let page;
let phantom;

phantomjs.create().then(function (instance) {
    phantom = instance;
    return phantom.createPage();
}).then(function (instance) {
    page = instance;
    return page.defineMethod('exists', function (selector, callback) {
        var element = this.evaluate(function (selector) {
            return document.querySelector(selector);
        }, selector);
        callback(element !== null);
    });
}).then(function (instance) {
    return page.open('http://google.com/');
}).then(function (status) {
    return page.invokeAsyncMethod('exists', 'input[type="submit"]');
}).then(function (exists) {
    console.log(exists);
    phantom.exit();
});
```

```js
'use strict';

const phantomjs = require('phantom');

let page;
let phantom;

phantomjs.create().then(function (instance) {
    phantom = instance;
    return phantom.createPage();
}).then(function (instance) {
    page = instance;
    return page.defineMethod('exists', function (selector) {
        var element = this.evaluate(function (selector) {
            return document.querySelector(selector);
        }, selector);
        return element !== null;
    });
}).then(function (instance) {
    return page.open('http://google.com/');
}).then(function (status) {
    return page.invokeMethod('exists', 'input[type="submit"]');
}).then(function (exists) {
    console.log(exists);
    phantom.exit();
});
```

#### Checklist
* [x] New tests have been added
* [x] `npm test` passes successfully
* [ ] Documentation has been updated

@amir20 to review

Hopefully I didn't mess anything up...

### CODE REVIEW

File: src/page.js [hunk 1]
```diff
@@ -43,17 +43,75 @@ export default class Page {
...
+     * Defines a method
+     */
+    defineMethod(name, implementation) {
+        return this.phantom.execute(this.target, 'defineMethod', [name, implementation]);
+    }
+    
+    /**
+     * Gets or sets a property
+     */
+    property() {
```
Comment (amir20) [inline src/page.js]: I am trying to figure out why `property` and `setting` got added. Can't they just be part of methods array?

File: src/shim.js [hunk 2]
```diff
@@ -97,7 +91,28 @@ const commands = {
         completeCommand(command);
     },
 
-    noop: command => completeCommand(command)
+    noop: command => completeCommand(command),
+    
+    invokeAsyncMethod: function (command) {
+        let target = objectSpace[command.target];
+        target[command.params[0]].apply(target, command.params.slice(1).concat(result => {
```
Comment (amir20) [inline src/shim.js]: Throw error here if target doesn't exist. This would never happen, but it would be good to say can't find target object to the user.

Comment (amir20) [inline src/shim.js]: Actually nevermind. I don't think we need it.

File: src/shim.js [hunk 3]
```diff
@@ -170,18 +185,8 @@ function executeCommand(command) {
...
-        } else {
-            let params = command.params.slice(); // copy params
-            params.push(status => {
-                command.response = status;
-                completeCommand(command);
-            });
-            method.apply(target, params);
-        }
+        command.response = method.apply(target, command.params);
+        completeCommand(command);
```
Comment (amir20) [inline src/shim.js]: Nice clean up!

**Comment by amir20** [comment]
Nice job! Everything looks good. I just don't know why you redefined `property` and `setting`? Once that's resolved, I can merge.

**Comment by elias-winberg** [comment]
`property`, and `setting` aren't members of the `WebPage` class, which means that they can't be invoked using `invokeAsyncMethod` or `invokeMethod`. I also thought it would be nice to distinguish between built-in methods and 'commands'.

We could change it to:

```js
// src/page.js

const commands = [
    'property',
    'setting'
];

commands.forEach(command => {
    Page.prototype[command] = function () {
        return this.phantom.execute(this.target, command, [].slice.call(arguments));
    }; 
});
```

**Comment by amir20** [comment]
OK that makes more sense. I went through the code and realized what was happening. Merging!

**Comment by amir20** [comment]
Fixes #487

File: src/shim.js (hunk 3 above)
Comment (elias-winberg) [inline src/shim.js]: Thanks! [The `else if` statement that follows](https://github.com/amir20/phantomjs-node/blob/master/src/shim.js#L185-L190) is in 'theory' deprecated, but the 'Page #reject(...) works when there is an error' test fails if the block is removed.

Comment (elias-winberg) [inline src/shim.js]: Never mind, I realized `phantom.exit` still relies  on the `else if` statement. Do you think we should bind `phantom.exit` to `invokeFunction`, and deprecate the `else if`?

**Pull request merged**